packages/runtime/plugin-garfish/src/runtime/utils/apps.tsx

```diff
@@ -48,25 +48,86 @@ function getAppInstance(
       React.ComponentType<any> | undefined
     >();
     const context = useContext(RuntimeReactContext);
-    const match = context?.router?.useRouteMatch?.();
-    const matchs = context?.router?.useMatches?.();
-    const location = context?.router?.useLocation?.();
-    let basename = options?.basename || '/';
-    if (matchs && matchs.length > 0) {
-      const matchItem = {
-        ...matchs[matchs.length - 1],
-      };
-      for (const key in matchItem.params) {
-        matchItem.pathname = matchItem.pathname.replace(
-          new RegExp(`/${matchItem.params[key]}$`),
-          '',
-        );
-      }
-      basename = pathJoin(basename, matchItem.pathname || '/');
-    } else if (match) {
-      basename = pathJoin(basename, match?.path || '/');
+
+    const useLocation = props.useLocation ?? context?.router?.useLocation;
+    const useHref = props.useHref ?? context?.router?.useHref;
+    const useHistory = props.useHistory ?? context?.router?.useHistory;
+    const isRouterV5 = Boolean(useHistory);
+
+    if (!useLocation) {
+      throw new Error(
+        `[@modern-js/plugin-garfish] Detected that the 'router: false' mode is used. In this case, the basename and popStateEvent cannot be correctly passed to the sub-app.
+You can manually pass 'useLocation' and 'useHref' props to assist plugin-garfish in calculating the "basename" and sync popStateEvent:
+if you are using react-router-V6:
+<Component useLocation={useLocation} useHref={useHref} />
+
+else react-router-V5:
+<Component useLocation={useLocation} useHistory={useHistory} />
+
+or directly pass the "basename":
+<Component basename={basename} useLocation={useLocation} />`,
+      );
     }
+    const location = useLocation();
+
+    let basename = '';
+    const activeWhen = appInfo.activeWhen as string;
 
+    // 1. use hooks to calculate the basename automatically
+    if (isRouterV5) {
+      // there is no dynamic switching of the router version in the project
+      // so hooks can be used in conditional judgment
+      // eslint-disable-next-line react-hooks/rules-of-hooks
+      const history = useHistory?.();
+      // To be compatible to history@4.10.1 and @5.3.0 we cannot write like this `history.createHref(pathname)`
+      basename = history?.createHref?.({ pathname: activeWhen });
+    } else {
+      // eslint-disable-next-line react-hooks/rules-of-hooks
+      basename = useHref?.(activeWhen);
+    }
+
+    // 2. options.basename in edenx.config.ts
+    /**
+     * e.g:
+    masterApp: {
+      basename: '/main-app-basename'
+    },
+     */
+    if (options?.basename && typeof options.basename === 'string') {
```

Comment:
This priority should be lower than the calculated basename

Reply:
It has been fixed